# 項目：評估並清理英國電商公司銷售數據

## 分析目標

此數據分析的目的是，根據市場銷售數據，挖掘`暢銷產品`，以便制定更有效的市場策略來提升營收。

本實戰計畫的目的在於練習評估資料乾淨和整潔度，並且基於評估結果，對資料進行清洗，從而得到可供下一步分析的資料。

## 簡介

原始資料集記錄了一家英國線上零售公司在2010年12月1日至2011年12月9日期間的所有交易情況，涵蓋了該公司在全球不同國家和地區的業務資料。

該公司主要銷售涵蓋各個場景的禮品，包括但不限於生日禮品、婚禮紀念品、聖誕禮品等等。該公司的客戶群主要包括批發商和個人消費者，其中批發商佔了相當大的比例。

資料每列的含義如下：
- `InvoiceNo`: 發票號碼。 6位數，作為交易的唯一識別碼。如果這個代碼以字母 `c` 開頭，表示這筆交易被`取消`。
- `StockCode`: 產品代碼。 5位數，作為產品的唯一識別碼。
- `Description`: 產品名稱。
- `Quantity`: 產品在交易中的數量。
- `InvoiceDate`: 發票日期和時間。交易發生的日期和時間。
- `UnitPrice`: 單價。價格單位為英鎊（£）。
- `CustomerID`: 客戶編號。 5位數，作為客戶的唯一識別碼。
- `Country`: 國家名稱。客戶所居住的國家的名稱。

## 讀取資料


In [2]:
import pandas as pd
original_data = pd.read_csv("e_commerce.csv")
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 評估資料

In [3]:
original_data.sample(10)

InvoiceNo StockCode                          Description  Quantity  \
361362    568346     22190                       LOCAL CAFE MUG         1   
229231    557020     22382           LUNCH BAG SPACEBOY DESIGN         10   
145824    548912     23194           GYMKHANA TREASURE BOOK BOX         8   
618       536412     21706  FOLDING UMBRELLA RED/WHITE POLKADOT         1   
429231    573548     21411          GINGHAM HEART  DOORSTOP RED         2   
228634    556936     82582            AREA PATROLLED METAL SIGN         4   
36734     539451     22717                   CARD DOG AND BALL          1   
223452    556484     22487       WHITE WOOD GARDEN PLANT LADDER         1   
337147    566431     23376  PACK OF 12 VINTAGE CHRISTMAS TISSUE        24   
502075    578834     21218               RED SPOTTY BISCUIT TIN         1   

             InvoiceDate  UnitPrice  CustomerID         Country  
361362   9/26/2011 15:28       2.46     14096.0  United Kingdom  
229231   6/16/2011 12:51       1.65     13246.0  United Kingdom  
145824     4/5/2011 9:32       2.25     16586.0  United Kingdom  
618      12/1/2010 11:49       4.95     17920.0  United Kingdom  
429231  10/31/2011 13:22       4.25     13680.0  United Kingdom  
228634   6/15/2011 16:04       2.10         NaN  United Kingdom  
36734   12/17/2010 16:59       0.85         NaN  United Kingdom  
223452   6/12/2011 13:17       9.95     16938.0  United Kingdom  
337147   9/12/2011 14:39       0.39     18180.0  United Kingdom  
502075  11/25/2011 15:32       3.75     18109.0  United Kingdom

In [4]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


總共有 `541909` 列

`Descriptoion` 以及 `CustomerID` 有缺失值

`InvoiceDate` 需為日期格式

`CustomerID` 需為字串格式

## 清理資料

In [5]:
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

`Description` 有 1454 行的缺失值

In [6]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

確認 `Description` 有缺失值時, `UnitPrice` 也都為 `0`

這些缺失值會影響產品分析，所以需要刪除

In [7]:
original_data[original_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

`CustomerID` 並不影響暢銷產品分析

### 評估重複資料
雖然 `InvoiceNo` `StockCode` `CustomerID` 都是唯一值，但一次交易可能包含多件商品，因此 `InvoiceNo` 可以包含重複值

不同交易可以包含同件商品，所以 `StockCode` 可以包含重複值

顧客可以進行多次交易或是購買多個商品，因此 `CustomerID` 也可以包含重複值

所以針對此筆資料，無須評估重複資料

### 評估不重複資料

In [8]:
original_data["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


`Country` 中， `USA` `United States`  均表示美國

`United Kingdom` `UK` `U.K` 均表示英國

需先對這些值進行統一

### 評估無效或錯誤資料

In [9]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

`CustomerID` 為客戶編號，應該要為字串

`Quantity` `UnitPrice` 有負數

In [10]:
original_data[original_data["Quantity"] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

`Quantity` 為負數時， `InvoiceNo` 似乎以 `C` 開頭，表示訂單被取消。

In [11]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

存在 `Quantity` 為負數且 `InvoiceNo` 不為 `C` 開頭的值

但是篩選出來的值 `UnitPrice` 為 `0` 因此增加 `UnitPrice` 的條件進行驗證

In [12]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C") & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

`Quantity` 為負數的時候， `InvoiceNo` 會為 `C` 開頭，或者 `UnitPrice` 為 `0`

因為會影響分析結果所以需要刪除

In [13]:
original_data[original_data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

篩選 `UnitPrice` 為負數 `Description` 的 `Adjust bad debt` 為壞帳調整

所以也需要刪除

## 清理資料

- `InvoiceDate` 需轉為日期格式
- `CustomerID` 需轉為字串格式
- `Description` 需刪除缺失值
- `Country` 將 `USA` 替換為 `United States` , 將 `UK` `U.K` 替換為 `United Kingdom`
- `Quantity` 將負數刪除
- `UnitPrice` 將負數刪除

In [15]:
cleaned_data = original_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

轉換 `InvoiceDate` 為日期格式

In [16]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"], format="%m/%d/%Y %H:%M")
cleaned_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

轉換 `CustomerID` 為字串

In [18]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

移除 `CustomerID` 結尾的 `.0`

In [21]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0, -2)
cleaned_data["CustomerID"]

0         178
1         178
2         178
3         178
4         178
         ... 
541904    126
541905    126
541906    126
541907    126
541908    126
Name: CustomerID, Length: 541909, dtype: object

刪除 `Description` 的缺失值

In [23]:
cleaned_data.dropna(subset=["Description"], inplace=True)
cleaned_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55        178  United Kingdom  
1      2010-12-01 08:26:00       3.39        178  United Kingdom  
2      2010-12-01 08:26:00       2.75        178  United Kingdom  
3      2010-12-01 08:26:00       3.39        178  United Kingdom  
4      2010-12-01 08:26:00       3.39        178  United Kingdom  
...                    ...        ...        ...             ...  
541904 2011-12-09 12:50:00       0.85        126          France  
541905 2011-12-09 12:50:00       2.10        126          France  
541906 2011-12-09 12:50:00       4.15        126          France  
541907 2011-12-09 12:50:00       4.15        126          France  
541908 2011-12-09 12:50:00       4.95        126          France  

[540455 rows x 8 columns]

確認 `Description` 是否還有缺失值

In [24]:
cleaned_data["Description"].isnull().sum()

0

替換 `Country` 的 `USA` 為 `United States`

In [25]:
cleaned_data["Country"] = cleaned_data["Country"].replace("USA", "United States")

檢查是否還有 `USA` 的值

In [26]:
len(cleaned_data[cleaned_data["Country"] == "USA"])

0

替換 `Country` 的 `UK` 以及 `U.K.` 為 `United Kingdom`

In [ ]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK": "Uited Kingdom", "U.K.": "United kingdom"})

檢查是否還有 `UK` 或 `U.K.` 的值

In [29]:
print(len(cleaned_data[cleaned_data["Country"] == "UK"]))
print(len(cleaned_data[cleaned_data["Country"] == "U.K."]))

0
0


刪除 `Quantity` 為負數的資料

In [30]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]

檢查是否還有 `Quantity` 為負數的值

In [31]:
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

刪除 `UnitPrice` 為負數的資料

In [33]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]

檢查是否還有 `UnitPrice` 為負數的值

In [34]:
len(cleaned_data[cleaned_data["UnitPrice"] < 0])

0

## 保存清理後的資料

In [36]:
cleaned_data.to_csv("e_commerce_cleaned.csv", index=False)

In [37]:
pd.read_csv("e_commerce_cleaned.csv").head()

C:\Users\coast\AppData\Local\Temp\ipykernel_33048\2686214642.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned.csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice  CustomerID         Country  
0  2010-12-01 08:26:00       2.55       178.0  United Kingdom  
1  2010-12-01 08:26:00       3.39       178.0  United Kingdom  
2  2010-12-01 08:26:00       2.75       178.0  United Kingdom  
3  2010-12-01 08:26:00       3.39       178.0  United Kingdom  
4  2010-12-01 08:26:00       3.39       178.0  United Kingdom